### ISSUE TITLE
Validate default values during model definition (not instantiation)

### ISSUE THREAD

**Issue by psychedelicious** [pydantic/pydantic#8722]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

Default values are only validated during model instantiation. I expect them to be validated during class definition.

```py
from pydantic.v1 import BaseModel as BaseModelV1, Field as FieldV1
from pydantic import BaseModel, Field
from typing import Literal

# No error when defining fields that are invalid by default
class MyModel(BaseModel):
    foo: Literal["bar"] = Field(default=123, validate_default=True)

# Same behaviour on V1
class MyModelV1(BaseModelV1):
    foo: Literal["bar"] = FieldV1(default=123)

    class Config:
        validate_all = True

# They throw on instantiation
MyModel()
MyModelV1()
```

In the example, it's clear that both models are invalid and will cause a runtime error if they are instantiated without `foo`. I expect to get an error when the module defining these classes is executed.

## Use case 1: accident insurance

Validating the defaults when the models are defined lets us catch accidents much sooner. The current behaviour means invalid defaults can lurk until we hit the right code path, then cause a runtime error at an unexpected time or for an unexpected reason.

`pyright` and `mypy` of course don't catch that the default value is typed incorrectly in these situations - we have to rely on runtime behaviour.

Tests could also catch this, though that could involve some really serious parameterization and feels like addressing the symptom not the root cause.

## Use-case 2: OpenAPI schema generation

When we provide invalid defaults, we kinda provide a booby-trapped contract to consumers of the API.

```py
class MyModel(BaseModel):
    foo: Literal["bar"] = Field(default=123, validate_default=True)

MyModel.model_json_schema()
# -->
# {
#   "properties": { "foo": { "const": "bar "default": 123, "title": "Foo" } },
#   "title": "MyModel",
#   "type": "object"
# }
```

This sends mixed messages. How am I supposed to interpret as an API consumer?

Two issues here, from an API consumer's point of view:
- It looks like this model wants `"bar"`, but will default to `123`. Does it also accept `int`? I don't want my `foo` to be stored as `123`!
- It looks like this model has a default value. In that case, I'll omit `foo` from my payload. Wait, why am I getting a 422 error?

Of course, this is also an issue without `validate_default` - you'd just not get a 422.

## Use case 2: user defined models

This is very much a nonstandard usage of pydantic, and there is an easy workaround. But I thought I'd describe the use-case anyways.

In our project, we provide a plugin-like API where we export some utilities that wrap `pydantic.Field`. Users define models which are imported into the app, and instantiated dynamically, in response to user actions.

I'd like to raise a runtime error when a model is defined with an invalid default value, and not need to wait until whatever situation arises that results in the runtime error.

This use-case has a simple workaround - we can validate the default in our wrapper.

## Desired API

The feature request adds a 3rd behaviour for `validate_default`, which suggests an enum for the arg:

```py
def Field(validate_default: Literal['never', 'definition', 'instantiation'] = 'never')
```

However, that's a breaking change, and I'm sure there are a lot of applications relying on the existing behaviour.

It could be a separate flag, but that would split the control of default validation across two flags, one of which would override the other. Feels messy.

Maybe:
```py
def Field(validate_default: bool | Literal['always'] = False)
```

Eh, also doesn't feel great.

## Summary

Overall this is a fairly minor issue

## Related issues

#1936 (unfortunately closed and missed the intent of the original request, which I interpret to be the same as mine)
#1280 (user was OK with validation during model instantiation, maintainer suggests existing behaviour is useful and good for perf)

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [X] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [ ] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by sydney-runkle** [comment]
I'll also add - I think we could add this in a minor release, given that we can avoid having this be a breaking change in any way. I'm going to remove the V3 tag.

**Comment by bjmc** [comment]
I'd like to see this change - we were just surprised by this. @psychedelicious have you made a start on an implementation? Any way I could help?

**Comment by psychedelicious** [comment]
@bjmc No, I have not started on this and don't anticipate working on it myself any time soon. I've never used rust beyond hello world - `pydantic-core` is pretty scary to me.

**Comment by mgroenbroek** [comment]
> @psychedelicious,
> 
> Thanks for the well thought out feature request. We think this would certainly be reasonable change to have, though for backwards compatability reasons we'd want to support both `str` and `bool` specifications for the `validate_defaults` flag.
> 
> PRs welcome to support this new behavior - it'll require changes both in `pydantic-core` and `pydantic`.

@sydney-runkle I may be interested in picking this up. Just so I understand correctly, you propose that the signature for `validate_default` on `Field` changes from:

```python
Field(
    ....,
    validate_default: bool | None = _Unset,
    ...
) -> Any
```

to

```python
Field(
    ....,
    validate_default: bool | str | None = _Unset,
    ...
) -> Any
```

Would you rather define a `Literal` or a string? We can then run the validator either as-is or based on the new behavior, where `never` has the same behavior as `False`, `instantiation` has the same behavior as `True` and `definition` is the new behavior.

**Comment by sydney-runkle** [comment]
@mgroenbroek let's go with `Literal`. Also, maybe `init` instead of `instantiation`

**Comment by mgroenbroek** [comment]
That’s a great suggestion. I have some time on my hands, so I am happy to pick this up.

**Comment by sydney-runkle** [comment]
We usually just keep one issue, but link it on both PRs.

> Do you tend to use make build-dev and point your pydantic-core dependency in pydantic/pyproject.toml to the generated wheel?

Exactly right!
